### ISSUE TITLE
"RuntimeError: dictionary changed size during iteration" in SSHConnection._cleanup()

### ISSUE THREAD

**Issue by thomasleveil** [ronf/asyncssh#315]
If running with Python 3, using `asyncssh.connect(...)` as a context manager, the _cleanup() call will raise `RuntimeError: dictionary changed size during iteration` if there are any local listeners to close.

This is due to a change in Python 3 regarding `Dict.values()` which now returns a [view object](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects) instead of a copy of the dictionary values [as in python 2](https://docs.python.org/2/library/stdtypes.html#dict.values)

Related code : https://github.com/ronf/asyncssh/blob/v2.4.2/asyncssh/connection.py#L584

the fix is to change :

```python
        for chan in list(self._channels.values()):
            chan.process_connection_close(exc)

        for listener in self._local_listeners.values():
            listener.close()
```

with :

```python
        for chan in list(self._channels.values()):
            chan.process_connection_close(exc)

        for listener in list(self._local_listeners.values()):
            listener.close()
```

This bug is present [since v0.2.0](https://github.com/ronf/asyncssh/commit/4be2b0ec90e3003c04b5ce2ab483f41e24d8b43d#diff-9e60bbfe4d79a23a44d680c5674f2b1eR149)

**Comment by ronf** [comment]
Thanks so much for the report and the suggested fix! Looking through the code, it appears that a similar fix is needed in a few other places. So far, I've identified:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index 497a3e3..4ec1888 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -581,7 +581,7 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         for chan in list(self._channels.values()):
             chan.process_connection_close(exc)
 
-        for listener in self._local_listeners.values():
+        for listener in list(self._local_listeners.values()):
             listener.close()
 
         while self._global_request_waiters:
@@ -2661,7 +2661,7 @@ class SSHClientConnection(SSHConnection):
             self._agent.close()
 
         if self._remote_listeners:
-            for listener in self._remote_listeners.values():
+            for listener in list(self._remote_listeners.values()):
                 listener.close()
 
             self._remote_listeners = {}
diff --git a/asyncssh/process.py b/asyncssh/process.py
index c2948e4..064d47d 100644
--- a/asyncssh/process.py
+++ b/asyncssh/process.py
@@ -776,10 +776,10 @@ class SSHProcess:
 
         super().connection_lost(exc)
 
-        for reader in self._readers.values():
+        for reader in list(self._readers.values()):
             reader.close()
 
-        for writer in self._writers.values():
+        for writer in list(self._writers.values()):
             writer.close()
 
         self._readers = {}
@@ -808,7 +808,7 @@ class SSHProcess:
 
         super().pause_writing()
 
-        for reader in self._readers.values():
+        for reader in list(self._readers.values()):
             reader.pause_reading()
 
     def resume_writing(self):
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index b85e925..7c7d573 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -1134,7 +1134,7 @@ class SFTPClientHandler(SFTPHandler):
 
         req_exc = exc or SFTPConnectionLost('Connection closed')
 
-        for waiter in self._requests.values():
+        for waiter in list(self._requests.values()):
             if not waiter.cancelled(): # pragma: no branch
                 waiter.set_exception(req_exc)
 
@@ -3321,7 +3321,7 @@ class SFTPServerHandler(SFTPHandler):
         """Clean up this SFTP server session"""
 
         if self._server: # pragma: no branch
-            for file_obj in self._file_handles.values():
+            for file_obj in list(self._file_handles.values()):
                 result = self._server.close(file_obj)
 
                 if inspect.isawaitable(result):
```

Many places in the code were already doing this, but not all. These fixes are now available as commit 89fd910 in the "develop" branch, and will be included in the next release.

**Comment by thomasleveil** [comment]
It's a pleasure to witness such reactivity from maintainers :cake: Many thanks for your work :heart:
